### DISCUSSION TITLE
Jog axis during hold/pause (M0)

### DISCUSSION THREAD

**Discussion by blackbmachine** [winder/Universal-G-Code-Sender#2572] [Help]
Ive read a couple other questions about this but Ive a unique application where it would be very helpful.

Using GRBL and UGS to run a CNC laser welder. It is set up as a rotary welder, with Y axis serving as the rotary, Z vertical distance to piece, and X axis moves to the location of each weld. Everything works great. The welder is designed to weld a stack of parts with 9-13 welds per unit.

Due to variation in some parts being welded the X location of some welds can vary up to about .005"- .010".

Ideally I would like to run the program and have the ability to have the machine move to its next weld location, Pause program, then jog the X axis by .005" (or whatever is necessary) and then hit start to run until the next M0 and repeat. 

There is no contact with the laser welding tip and work so tool offsets are irrelevant.

Example:

G0 X1.85 Y-0.069 Z.5
G1 Z0 F30
M00 (jog to exact location *say right X+ .005" then hit start)
M8
G4 P3
M3 S1000

EDIT: in this case the actual weld would lay in at X1.855. The program then continues with the original (very close) X values to be individually modified after each M0 command

**Comment by breiler** [comment]
I believe that this is really hard to solve in UGS. During "M0" hold the controller won't accept other commands such as jog commands and/or work offset commands.

To do this we would need to implement a special macro command in UGS which will require a lot of work.

Is it the same program running in a loop? In that case have you considered making the program so that it moves to the next location at the end of the program then fine adjust with jog and then run the program again?

**Reply by blackbmachine** [reply]
Unfortunately each weld has a different start/stop point on the rotary (Y) axis relative to the other welds. Also, each weld is done in a specific order to prevent weld deflection as much as possible. I did try each of the welds as a separate program with each X manually jogged into place and it does work fine. The issue is it takes the unit weld time from just over 2 minutes to over 20. It also introduces a significant potential for error. Due to the nature of these parts, errors are very expensive. 

I guess what I am hoping for is more than likely not possible without some sort of source code modification that would be pretty dangerous for the typical use of UGS. I would attempt this "special macro command" but I wouldnt even know where to start. Thank You.

**Reply by blackbmachine** [reply]
I didnt think through your suggestion thoroughly enough. I was ending the individual programs after the weld then jogging to location and starting again. By putting the next move into each program after the weld, I could up the efficiency by about half and decrease potential for errors significantly vs 13 very slightly different programs run in a specific order.

Ill play with this idea as a temporary fix at least. Thanks
